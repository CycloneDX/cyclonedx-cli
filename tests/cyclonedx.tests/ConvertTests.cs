// This file is part of CycloneDX CLI Tool
//
// Licensed under the Apache License, Version 2.0 (the “License”);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an “AS IS” BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) OWASP Foundation. All Rights Reserved.
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Xunit;
using Snapshooter;
using Snapshooter.Xunit;
using CycloneDX.Cli.Commands;

namespace CycloneDX.Cli.Tests
{
    public class ConvertTests
    {
        [Theory]
        [InlineData("bom-1.0.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.0.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.0.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_1)]

        [InlineData("bom-1.1.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.1.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.1.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_1)]

        [InlineData("bom-1.2.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.2.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.2.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, null)]
        [InlineData("bom-1.2.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_2)]

        [InlineData("bom-1.2.json", ConvertFormat.autodetect, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.2.json", ConvertFormat.json, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.2.json", ConvertFormat.json, "bom.json", ConvertFormat.json, SpecificationVersion.v1_2)]

        [InlineData("bom-1.3.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.3.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.3.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, null)]
        [InlineData("bom-1.3.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_3)]

        [InlineData("bom-1.3.json", ConvertFormat.autodetect, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.3.json", ConvertFormat.json, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.3.json", ConvertFormat.json, "bom.json", ConvertFormat.json, null)]
        [InlineData("bom-1.3.json", ConvertFormat.json, "bom.json", ConvertFormat.json, SpecificationVersion.v1_3)]

        [InlineData("bom-1.4.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.4.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.4.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, null)]
        [InlineData("bom-1.4.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_4)]

        [InlineData("bom-1.5.xml", ConvertFormat.autodetect, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.5.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.5.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, null)]
        [InlineData("bom-1.5.xml", ConvertFormat.xml, "bom.xml", ConvertFormat.xml, SpecificationVersion.v1_5)]

        [InlineData("bom-1.4.json", ConvertFormat.autodetect, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.4.json", ConvertFormat.json, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.4.json", ConvertFormat.json, "bom.json", ConvertFormat.json, null)]
        [InlineData("bom-1.4.json", ConvertFormat.json, "bom.json", ConvertFormat.json, SpecificationVersion.v1_4)]

        [InlineData("bom-1.5.json", ConvertFormat.autodetect, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.5.json", ConvertFormat.json, "bom.json", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.5.json", ConvertFormat.json, "bom.json", ConvertFormat.json, null)]
        [InlineData("bom-1.5.json", ConvertFormat.json, "bom.json", ConvertFormat.json, SpecificationVersion.v1_5)]

        [InlineData("bom.csv", ConvertFormat.autodetect, "bom.csv", ConvertFormat.autodetect, null)]
        [InlineData("bom.csv", ConvertFormat.csv, "bom.csv", ConvertFormat.autodetect, null)]
        [InlineData("bom.csv", ConvertFormat.csv, "bom.csv", ConvertFormat.csv, null)]

        [InlineData("bom-1.3.cdx", ConvertFormat.protobuf, "bom.cdx", ConvertFormat.autodetect, null)]
        [InlineData("bom-1.3.cdx", ConvertFormat.protobuf, "bom.json", ConvertFormat.json, SpecificationVersion.v1_3)]
        [InlineData("bom-1.3.json", ConvertFormat.json, "bom.cdx", ConvertFormat.protobuf, SpecificationVersion.v1_3)]
        public async Task Convert(string inputFilename, ConvertFormat inputFormat, string outputFilename, ConvertFormat outputFormat, SpecificationVersion? outputVersion)
        {
            using (var tempDirectory = new TempDirectory())
            {
                var fullOutputPath = Path.Join(tempDirectory.DirectoryPath, outputFilename);
                var exitCode = await ConvertCommand.Convert(new ConvertCommandOptions
                {
                    InputFile = Path.Combine("Resources", inputFilename),
                    OutputFile = fullOutputPath,
                    InputFormat = inputFormat,
                    OutputFormat = outputFormat,
                    OutputVersion = outputVersion,
                }).ConfigureAwait(false);
                
                Assert.Equal(0, exitCode);
                var bom = File.ReadAllText(fullOutputPath);
                Snapshot.Match(bom, SnapshotNameExtension.Create(inputFilename, inputFormat, outputFilename, outputFormat, outputVersion));
            }
        }

        [Theory]
        [InlineData(ConvertFormat.autodetect)]
        [InlineData(ConvertFormat.spdxjson)]
        public async Task ConvertToSpdxJson(ConvertFormat outputFormat)
        {
            using (var tempDirectory = new TempDirectory())
            {
                var outputFilename = Path.Combine(tempDirectory.DirectoryPath, "bom.spdx.json");
                var exitCode = await ConvertCommand.Convert(new ConvertCommandOptions
                {
                    InputFile = Path.Combine("Resources", "document.spdx.json"),
                    OutputFile = outputFilename,
                    InputFormat = ConvertFormat.autodetect,
                    OutputFormat = outputFormat
                    
                }).ConfigureAwait(false);
                
                Assert.Equal(0, exitCode);
                var bom = File.ReadAllText(outputFilename);
                bom = Regex.Replace(bom, @"Created"": .*\n", "");
                Snapshot.Match(bom, SnapshotNameExtension.Create(outputFormat));
            }
        }

        [Theory]
        [InlineData(ConvertFormat.autodetect)]
        [InlineData(ConvertFormat.spdxjson)]
        public async Task ConvertFromSpdxJson(ConvertFormat inputFormat)
        {
            using (var tempDirectory = new TempDirectory())
            {
                var outputFilename = Path.Combine(tempDirectory.DirectoryPath, "bom.spdx.json");
                var exitCode = await ConvertCommand.Convert(new ConvertCommandOptions
                {
                    InputFile = Path.Combine("Resources", "document.spdx.json"),
                    OutputFile = outputFilename,
                    InputFormat = inputFormat,
                    OutputFormat = ConvertFormat.xml,
                    
                }).ConfigureAwait(false);
                
                Assert.Equal(0, exitCode);
                var bom = File.ReadAllText(outputFilename);
                bom = Regex.Replace(bom, @"Created"": .*\n", "");
                Snapshot.Match(bom, SnapshotNameExtension.Create(inputFormat));
            }
        }
    }
}
