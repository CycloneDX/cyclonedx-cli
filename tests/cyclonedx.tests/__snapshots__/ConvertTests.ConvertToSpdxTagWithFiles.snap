SPDXVersion: SPDX-2.2
DataLicense: CC0-1.0
SPDXID: SPDXRef-DOCUMENT
DocumentName: Acme Application-9.1.1
DocumentNamespace: http://spdx.org/spdxdocs/Acme Application-9.1.1-3e671687-395b-41f5-a30f-a58921a69b79
LicenseListVersion: 3.14
Creator: Person: Samantha Wright (samantha.wright@example.com)
Creator: Tool: Awesome Tool-9.1.2
Creator: Tool: CycloneDX-CLI
CreatorComment: <text>This SPDX document was created by conversion from a CycloneDX BOM.</text>
DocumentComment: <text>SPDX does not support all features of CycloneDX. Some information is missing from the original CycloneDX BOM.</text>

PackageName: com.acme tomcat-catalina
SPDXID: SPDXRef-pkg-npm-acme-component-1.0.0
PackageVersion: 9.0.14
PackageOriginator: Person: Acme Inc ()
PackageDownloadLocation: NOASSERTION
FilesAnalyzed: false
PackageChecksum: MD5: 3942447fac867ae5cdb3229b658f4d48
PackageChecksum: SHA1: e6b1000b94e835ffd37f4c6dcbdad43f4b48a02a
PackageChecksum: SHA256: f498a8ff2dd007e29c2074f5e4b01a9a01775c3ff3aeaf6906ea503bc5791b7b
PackageChecksum: SHA512: e8f33e424f3f4ed6db76a482fde1a5298970e442c531729119e37991884bdffab4f9426b7ee11fccd074eeda0634d71697d6f88a460dce0ac8d627a29f7d1282
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: Apache-2.0
PackageCopyrightText: NOASSERTION
ExternalRef: PACKAGE-MANAGER purl pkg:npm/acme/component@1.0.0

FileName: somefile.dll
SPDXID: SPDXRef-2
FileLicenseConcluded: NOASSERTION
LicenseInfoInFile: NOASSERTION
FileContributor: Example Super Heros
